pkg/graveler/committed/prefix_iterator.go

```diff
@@ -0,0 +1,166 @@
+package committed
+
+import (
+	"sort"
+	"strings"
+
+	"github.com/treeverse/lakefs/pkg/graveler"
+)
+
+type rangeValue struct {
+	r  *Range
+	vr *graveler.ValueRecord
+}
+
+type ImportIterator interface {
+	IsCurrentRangeBoundedByPath() bool
+	IsCurrentPathIncludedInRange() bool
+}
+
+type PrefixIterator struct {
+	currentPrefixIndex int
+	prefixes           []graveler.Prefix
+	rangeIterator      Iterator
+	currentRangeValue  rangeValue
+}
+
+func (ipi *PrefixIterator) Value() (*graveler.ValueRecord, *Range) {
+	return ipi.currentRangeValue.vr, ipi.currentRangeValue.r
+}
+
+func (ipi *PrefixIterator) updateValue() (*graveler.ValueRecord, *Range) {
+	vr, r := ipi.rangeIterator.Value()
+	ipi.currentRangeValue = rangeValue{
+		r,
+		vr,
+	}
+	return vr, r
+}
+
+func (ipi *PrefixIterator) Err() error {
+	return ipi.rangeIterator.Err()
+}
+func (ipi *PrefixIterator) Close() {
+	ipi.rangeIterator.Close()
+}
+func (ipi *PrefixIterator) SeekGE(id graveler.Key) {
+	ipi.rangeIterator.SeekGE(id)
+}
+
+func (ipi *PrefixIterator) Next() bool {
+	hasNext := ipi.rangeIterator.Next()
+	if !hasNext {
+		return false
+	}
+	vr, r := ipi.updateValue()
+	ipi.updatePath()
+
+	if vr == nil && r != nil { // head of range
+		for ipi.IsCurrentRangeBoundedByPath() {
```

Comment:
I don't understand this loop: it runs as long as the current range has the right prefix!  I would expect to call `ipi.rangeIterator.NextRange()` _once_, then loop while `!ipi.IsCurrentRangeBoundedByPath()`.

Reply:
get to the first range that isn't bounded by the current prefix, which is the next range to be merged